README.md
diceware
========

**tl;dr** diceware is an implementation of the [diceware passphrase](http://world.std.com/~reinhold/diceware.html), created by Arnold G. Reinhold

[![Build Status](https://drone.io/github.com/jbrodriguez/diceware/status.png)](https://drone.io/github.com/jbrodriguez/diceware/latest)

## Screen output
```
$ ./diceware -e -s -w 8
diceware v0.1.0-1.9fdd687
Copyright (c) 2015 Juan B. Rodriguez
Web: http://jbrodriguez.io
Github: https://github.com/jbrodriguez/diceware

Using the default word list ...

You chose to create a password of 8 words, which is ~103 bits of entropy (~12.92 bits/word and ~5.16/symbol).
That many words equates to an effective keyspace of ~6,683,747,269,421,867,000,000,000,000,000 possible keys,
(7776^words)/2 which is 50% of the total key space since on average a key will be found in that reduced keyspace.
Conservatively assuming an adversary can guess passwords at the rate of a trillion (1,000,000,000,000) keys/second,
as Edward Snowden claims, an exhaustive brute-force search on half the total keyspace might take:

~6,683,747,269,421,867,000 seconds
~111,395,787,823,697,780 minutes
~1,856,596,463,728,296 hours
~77,358,185,988,679 days
~211,940,235,585 years
~3,153,872,553 x avg. lifespan
~211,940,236 millenia
~15 x age Universe

Learn more about calculating entropy at http://world.std.com/~reinhold/dicewarefaq.html#calculatingentropy

Roll 5 dices 8 times and enter the result after each prompt (press Ctrl+C to stop at any moment):

Roll #1: 64122
Roll #2: 56666
Roll #3: 65546
Roll #4: 25231
Roll #5: 41321
Roll #6: 23656
Roll #7: 21361
Roll #8: 61243

Your password is:
whim the 44 far lw dyke coot tift


Entropy: 130.241000
Crack Time: 80434485034394646894966093304037376.000000 seconds
Alternate Crack Time: centuries
Score: 4

Key for score:
0 = too guessable: risky password
1 = very guessable: protection from throttled online attacks
2 = somewhat guessable: protection from unthrottled online attacks
3 = safely unguessable: moderate protection from offline slow-hash scenario
4 = very unguessable: strong protection from offline slow-hash scenario
```

## Rationale
There's already a great diceware golang version in https://github.com/natefinch/diceware.

But I happen to own a set of Bicycle dice, which although are not the same as Casino dice, they're still quite usable.

So I created a utility that allows me to enter the result of my rolling the dice, and output the  password.

Additionally, I added some stats:
- Some theoretical entropy stats based on number of words in the password (taken from https://grempe.github.io/diceware/)
- A password strength meter based on [dropbox's zxcvbn implementation](https://github.com/dropbox/zxcvbn) (and golang's version in https://github.com/nbutton23/zxcvbn-go)

## Installation
Download the binary that corresponds to your platform.

## Development
It's assumed you have GO15VENDOREXPERIMENT set to 1.

The recommended way is to install [gom](https://github.com/mattn/gom), then

```
$ git clone https://github.com/jbrodriguez/diceware
$ cd diceware
$ gom install
$ make build
$ ./diceware
```

Alternatively (but you lose vendor dependency management), you can

```
$ go get https://github.com/nbutton23/zxcvbn-go
$ go get github.com/stretchr/testify(for testing only)
$ git clone https://github.com/jbrodriguez/diceware
$ cd diceware
$ make build
$ ./diceware
```

## Usage
```
$ ./diceware -h
Usage of ./diceware:
  -b    use alternate word list from Alan Beale
  -e    display entropy
  -s    display password strength
  -w int
        number of words to generate (default 6)
```

## Credits
- [Arnold G. Reinhold](http://world.std.com/~reinhold/diceware.html)
- [Glenn Rempe](https://github.com/grempe/diceware)
- [Nate Finch](https://github.com/natefinch/diceware)
- [Nathan Button](https://github.com/natefinch/diceware)

## LICENSE
MIT


CHANGES
Version 0.1.4 - 2015-11-22
    - Add drone.io badge


Version 0.1.3 - 2015-11-22
    - Add missing words file
    - Fix readme copy


Version 0.1.2 - 2015-11-22
    - Add readme


Version 0.1.1 - 2015-11-22
    - Add some tests


Version 0.1.0 - 2015-11-22
    Initial commit




diceware_test.go
package main

import (
	// "flag"
	// "fmt"
	"github.com/stretchr/testify/assert"
	"strings"
	"testing"
)

var rolls, expectedDefault, expectedBeale []string

func TestMain(m *testing.M) {
	rolls = []string{
		"64122",
		"56666",
		"65546",
		"25231",
		"41321",
		"23656",
		"21361",
		"61243",
	}

	expectedDefault = []string{
		"whim",
		"the",
		"44",
		"far",
		"lw",
		"dyke",
		"coot",
		"tift",
	}

	expectedBeale = []string{
		"wm",
		"tile",
		"222",
		"faze",
		"madam",
		"ears",
		"coral",
		"toni",
	}

	m.Run()
}

func TestGetPassword(t *testing.T) {
	chosenDefault := getPassword(strings.NewReader(list), rolls)
	chosenBeale := getPassword(strings.NewReader(bealeList), rolls)

	assert.Equal(t, expectedDefault, chosenDefault, "expectedDefault is wrong")
	assert.Equal(t, expectedBeale, chosenBeale, "expectedBeale is wrong")
}

func TestStrength(t *testing.T) {
	strengthDefault := getPasswordStrength(strings.Join(expectedDefault, " "), nil)
	strengthBeale := getPasswordStrength(strings.Join(expectedBeale, " "), nil)

	// Entropy
	assert.Equal(t, 130.241, strengthDefault.Entropy, "strengthDefault.Entropy is wrong")
	assert.Equal(t, 124.698, strengthBeale.Entropy, "strengthBeale.Entropy is wrong")

	// CrackTime
	assert.Equal(t, 8.043448503439465e+34, strengthDefault.CrackTime, "strengthDefault.CrackTime is wrong")
	assert.Equal(t, 1.725027324318945e+33, strengthBeale.CrackTime, "strengthBeale.CrackTime is wrong")

	// CrackTimeDisplay
	assert.Equal(t, "centuries", strengthDefault.CrackTimeDisplay, "strengthDefault.CrackTimeDispla is wrong")
	assert.Equal(t, "centuries", strengthBeale.CrackTimeDisplay, "strengthBeale.CrackTimeDisplay is wrong")

	// Score
	assert.Equal(t, 4, strengthDefault.Score, "strengthDefault.Score is wrong")
	assert.Equal(t, 4, strengthBeale.Score, "strengthBeale.Score is wrong")
}

func TestWordsEntropy(t *testing.T) {
	entropy := getWordsEntropy(8, 1000000000000.)

	assert.Equal(t, 103.36, entropy.Entropy, "entropy.Entropy is wrong")
	assert.Equal(t, 6.683747269421867e+30, entropy.HalfKeySpace, "entropy.HalfKeySpace is wrong")
	assert.Equal(t, 6.683747269421867e+18, entropy.Seconds, "entropy.Seconds is wrong")
	assert.Equal(t, 1.1139578782369778e+17, entropy.Minutes, "entropy.Minutes is wrong")
	assert.Equal(t, 1.8565964637282962e+15, entropy.Hours, "entropy.Hours is wrong")
	assert.Equal(t, 7.735818598867902e+13, entropy.Days, "entropy.Days is wrong")
	assert.Equal(t, 2.1194023558542197e+11, entropy.Years, "entropy.Years is wrong")
	assert.Equal(t, 3.1538725533544936e+09, entropy.HumanLifetimes, "entropy.HumanLifetimes is wrong")
	assert.Equal(t, 2.1194023558542198e+08, entropy.Millenia, "entropy.Millenia is wrong")
	assert.Equal(t, 15.360214203900707, entropy.UniverseLifetimes, "entropy.UniverseLifetimes is wrong")
}


diceware.go
package main

import (
	"bytes"
	"encoding/csv"
	// "encoding/json"
	"flag"
	"fmt"
	"github.com/nbutton23/zxcvbn-go/matching"
	"github.com/nbutton23/zxcvbn-go/scoring"
	"github.com/nbutton23/zxcvbn-go/utils/math"
	"io"
	"log"
	"math"
	// "os"
	"strconv"
	"strings"
	"time"
)

var Version string

const diceCount = 5

type Entropy struct {
	Entropy           float64
	KeySpace          float64
	HalfKeySpace      float64
	Seconds           float64
	Minutes           float64
	Hours             float64
	Days              float64
	Years             float64
	Millenia          float64
	HumanLifetimes    float64
	UniverseLifetimes float64
}

func main() {
	var length int
	var beale bool
	var entropy bool
	var strength bool

	flag.IntVar(&length, "w", 6, "number of words to generate")
	// flag.BoolVar(&extra, "e", false, "generate an extra random character")
	flag.BoolVar(&beale, "b", false, "use alternate word list from Alan Beale")
	flag.BoolVar(&entropy, "e", false, "display entropy")
	flag.BoolVar(&strength, "s", false, "display password strength")
	flag.Parse()

	fmt.Printf("diceware v%s\n", Version)
	fmt.Println("Copyright (c) 2015 Juan B. Rodriguez")
	fmt.Println("Web: http://jbrodriguez.io")
	fmt.Println("Github: https://github.com/jbrodriguez/diceware")
	fmt.Println()

	wordList := "default"
	if beale {
		wordList = "alternate Alan Beale's"
	}

	fmt.Printf("Using the %s word list ...\n", wordList)

	if entropy {
		displayEntropy(length)
	}

	fmt.Println()
	fmt.Printf("Roll 5 dices %d times and enter the result after each prompt (press Ctrl+C to stop at any moment):\n\n", length)

	rolls := make([]string, length)
	count := 1

	for {
		if count > length {
			break
		}

		var ln string
		fmt.Printf("Roll #%d: ", count)
		fmt.Scanf("%s", &ln)

		if len(ln) != diceCount {
			fmt.Printf("You must input exactly %d digits. Please try again\n", diceCount)
			continue
		}

		if !digitsAreOk(ln, diceCount) {
			fmt.Printf("Each digit must be between 1 and 6. Please try again\n")
			continue
		}

		rolls[count-1] = ln
		count++
	}

	var chosen []string
	if beale {
		chosen = getPassword(strings.NewReader(bealeList), rolls)
	} else {
		chosen = getPassword(strings.NewReader(list), rolls)
	}

	fmt.Println("\nYour password is:")
	fmt.Printf("%s\n", strings.Join(chosen, " "))

	if strength {
		fmt.Println()
		displayStrength(strings.Join(chosen, " "))
	}

}

func digitsAreOk(line string, count int) (ok bool) {
	for i := 0; i < count; i++ {
		digit, err := strconv.Atoi(string(line[i]))
		ok = err == nil && digit >= 1 && digit <= 6
		if !ok {
			break
		}
	}

	return ok
}

func getPassword(rd io.Reader, rolls []string) (chosen []string) {
	r := csv.NewReader(rd)
	r.Comma = '\t'
	r.FieldsPerRecord = 2

	words := make(map[string]string, 7776)
	rec, err := r.Read()
	for err == nil {
		words[rec[0]] = rec[1]
		rec, err = r.Read()
	}

	if err != io.EOF {
		log.Fatalln(err)
	}

	if len(words) != 7776 {
		log.Fatalf("Expected exactly 7776 words in tsv wordlist, but found %d", len(words))
	}

	chosen = make([]string, len(rolls))
	for k, v := range rolls {
		chosen[k] = words[v]
	}

	return chosen
}

func getWordsEntropy(numWords int, guessesPerSec float64) Entropy {
	entropy := Entropy{}

	var (
		avgHumanLifespanInYears float64 = 67.2         // https://en.wikipedia.org/wiki/Life_expectancy
		ageOfUniverseInYears    float64 = 13798000000. // https://en.wikipedia.org/wiki/Age_of_the_universe
	)

	entropy.Entropy = 12.92 * float64(numWords)

	// https://xkcd.com/936/
	// https://security.stackexchange.com/questions/62832/is-the-oft-cited-xkcd-scheme-no-longer-good-advice/62881#62881
	// https://hashcat.net/forum/thread-2580.html
	entropy.KeySpace = math.Pow(7776., float64(numWords))

	// Divide the keySpace in half. On average it is expected that an
	// exhaustive search of only half the keySpace will result in success.
	entropy.HalfKeySpace = entropy.KeySpace / 2

	// // "Assume that your adversary is capable of a trillion guesses per second" - Snowden
	// // http://www.nytimes.com/2013/08/18/magazine/laura-poitras-snowden.html?pagewanted=all&_r=0
	// guessesPerSec = 1000000000000.

	entropy.Seconds = entropy.HalfKeySpace / guessesPerSec
	entropy.Minutes = entropy.Seconds / 60
	entropy.Hours = entropy.Minutes / 60
	entropy.Days = entropy.Hours / 24
	entropy.Years = entropy.Days / 365
	entropy.Millenia = entropy.Years / 1000
	entropy.HumanLifetimes = entropy.Years / avgHumanLifespanInYears
	entropy.UniverseLifetimes = entropy.Years / ageOfUniverseInYears

	return entropy
}

func getPasswordStrength(password string, userInputs []string) scoring.MinEntropyMatch {
	start := time.Now()
	matches := matching.Omnimatch(password, userInputs)
	result := scoring.MinimumEntropyMatchSequence(password, matches)
	end := time.Now()

	calcTime := end.Nanosecond() - start.Nanosecond()
	result.CalcTime = zxcvbn_math.Round(float64(calcTime)*time.Nanosecond.Seconds(), .5, 3)
	return result
}

func displayEntropy(numWords int) {
	entropy := getWordsEntropy(numWords, 1000000000000.)

	fmt.Printf(`
You chose to create a password of %d words, which is ~%s bits of entropy (~12.92 bits/word and ~5.16/symbol).
That many words equates to an effective keyspace of ~%s possible keys,
(7776^words)/2 which is 50%% of the total key space since on average a key will be found in that reduced keyspace.
Conservatively assuming an adversary can guess passwords at the rate of a trillion (1,000,000,000,000) keys/second,
as Edward Snowden claims, an exhaustive brute-force search on half the total keyspace might take:

~%s seconds
~%s minutes
~%s hours
~%s days
~%s years
~%s x avg. lifespan
~%s millenia
~%s x age Universe

Learn more about calculating entropy at http://world.std.com/~reinhold/dicewarefaq.html#calculatingentropy	
`,
		numWords,
		Commaf(trunc(entropy.Entropy)),
		Commaf(trunc(entropy.HalfKeySpace)),
		Commaf(trunc(entropy.Seconds)),
		Commaf(trunc(entropy.Minutes)),
		Commaf(trunc(entropy.Hours)),
		Commaf(trunc(entropy.Days)),
		Commaf(trunc(entropy.Years)),
		Commaf(trunc(entropy.HumanLifetimes)),
		Commaf(trunc(entropy.Millenia)),
		Commaf(trunc(entropy.UniverseLifetimes)),
	)

}

func displayStrength(password string) {
	strength := getPasswordStrength(password, nil)

	fmt.Printf(`
Entropy: %f
Crack Time: %f seconds
Alternate Crack Time: %s
Score: %d

Key for score:
0 = too guessable: risky password
1 = very guessable: protection from throttled online attacks
2 = somewhat guessable: protection from unthrottled online attacks
3 = safely unguessable: moderate protection from offline slow-hash scenario
4 = very unguessable: strong protection from offline slow-hash scenario
`,
		strength.Entropy,
		strength.CrackTime,
		strength.CrackTimeDisplay,
		strength.Score,
	)
}

// b, err := json.MarshalIndent(strength, "", " ")
// if err != nil {
// 	fmt.Println(err)
// 	os.Exit(1)
// }

// fmt.Println(string(b))

func Commaf(v float64) string {
	buf := &bytes.Buffer{}
	if v < 0 {
		buf.Write([]byte{'-'})
		v = 0 - v
	}

	comma := []byte{','}

	parts := strings.Split(strconv.FormatFloat(v, 'f', -1, 64), ".")
	pos := 0
	if len(parts[0])%3 != 0 {
		pos += len(parts[0]) % 3
		buf.WriteString(parts[0][:pos])
		buf.Write(comma)
	}
	for ; pos < len(parts[0]); pos += 3 {
		buf.WriteString(parts[0][pos : pos+3])
		buf.Write(comma)
	}
	buf.Truncate(buf.Len() - 1)

	if len(parts) > 1 {
		buf.Write([]byte{'.'})
		buf.WriteString(parts[1])
	}
	return buf.String()
}

func trunc(v float64) float64 {
	if v < 1 {
		return v
	} else {
		return rint(v)
	}
}

func rint(x float64) float64 {
	v, frac := math.Modf(x)
	if x > 0.0 {
		if frac > 0.5 || (frac == 0.5 && uint64(v)%2 != 0) {
			v += 1.0
		}
	} else {
		if frac < -0.5 || (frac == -0.5 && uint64(v)%2 != 0) {
			v -= 1.0
		}
	}

	return v
}
